flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/util/RetryableExceptionClassifier.java

```
@@ -0,0 +1,90 @@
+    private RetryableExceptionClassifier chainedClassifier;
+
+    public RetryableExceptionClassifier(
+            Predicate<Throwable> validator, Function<Throwable, Exception> throwableMapper) {
+        this.throwableMapper = throwableMapper;
+        this.validator = validator;
+        this.chainedClassifier = null;
+    }
+
+    public boolean shouldSuppress(Throwable err, Consumer<Exception> throwableConsumer) {
+        if (validator.test(err)) {
+            throwableConsumer.accept(throwableMapper.apply(err));
+            return false;
+        }
+
+        if (chainedClassifier != null) {
+            return chainedClassifier.shouldSuppress(err, throwableConsumer);
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public RetryableExceptionClassifier clone() {
+        return new RetryableExceptionClassifier(validator, throwableMapper);
+    }
+
+    public static RetryableExceptionClassifier withRootCauseOfType(
+            Class<? extends Throwable> type, Function<Throwable, Exception> mapper) {
+        return new RetryableExceptionClassifier(
+                err -> ExceptionUtils.findThrowable(err, type).isPresent(), mapper);
+    }
+
+    public static RetryableExceptionClassifier withRootCauseWithMessage(
+            String message, Function<Throwable, Exception> mapper) {
+        return new RetryableExceptionClassifier(
+                err -> ExceptionUtils.findThrowableWithMessage(err, message).isPresent(), mapper);
+    }
+
+    public static RetryableExceptionClassifier createChain(
+            RetryableExceptionClassifier... classifiers) {
+        Set<RetryableExceptionClassifier> importedWrappers = new HashSet<>();
+        RetryableExceptionClassifier taleClassifier = classifiers[0];
+        importedWrappers.add(taleClassifier);
+        for (int i = 1; i < classifiers.length; ++i) {
+            if (importedWrappers.contains(classifiers[i])) {
+                continue;
```

I think we should throw an `InvalidArgumentException` here